compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -41,8 +41,11 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
   def Super(qual: Tree, mixName: TypeName, inConstrCall: Boolean, mixinClass: Symbol = NoSymbol)(implicit ctx: Context): Super =
     Super(qual, if (mixName.isEmpty) untpd.EmptyTypeIdent else untpd.Ident(mixName), inConstrCall, mixinClass)
 
-  def Apply(fn: Tree, args: List[Tree])(implicit ctx: Context): Apply =
+  def Apply(fn: Tree, args: List[Tree])(implicit ctx: Context): Apply = {
+    assert(!fn.isInstanceOf[Block])
+    assert(!fn.isInstanceOf[If])
     ta.assignType(untpd.Apply(fn, args), fn, args)
+  }
 
   def TypeApply(fn: Tree, args: List[Tree])(implicit ctx: Context): TypeApply =
     ta.assignType(untpd.TypeApply(fn, args), fn, args)
```

Comment:
You could try to do the same for TypeApply.

Reply:
Yes, we could do that. I'm wondering how `if/else` could possibly appear in `fun`. But it is always good to have these invariants if they don't incur much performance cost.